In [217]:
import pandas as pd
import numpy as np


In [218]:
x_test_final = pd.read_csv('/content/drive/MyDrive/Ensemble/X_test_final.csv')
x_train = pd.read_csv('/content/drive/MyDrive/Ensemble/X_train_NHkHMNU.csv')
y_test_random_final = pd.read_csv('/content/drive/MyDrive/Ensemble/y_test_random_final.csv')
y_train = pd.read_csv('/content/drive/MyDrive/Ensemble/y_train_ZAN5mwg.csv')

In [219]:
x_test_final.columns

Index(['ID', 'DAY_ID', 'COUNTRY', 'DE_CONSUMPTION', 'FR_CONSUMPTION',
       'DE_FR_EXCHANGE', 'FR_DE_EXCHANGE', 'DE_NET_EXPORT', 'FR_NET_EXPORT',
       'DE_NET_IMPORT', 'FR_NET_IMPORT', 'DE_GAS', 'FR_GAS', 'DE_COAL',
       'FR_COAL', 'DE_HYDRO', 'FR_HYDRO', 'DE_NUCLEAR', 'FR_NUCLEAR',
       'DE_SOLAR', 'FR_SOLAR', 'DE_WINDPOW', 'FR_WINDPOW', 'DE_LIGNITE',
       'DE_RESIDUAL_LOAD', 'FR_RESIDUAL_LOAD', 'DE_RAIN', 'FR_RAIN', 'DE_WIND',
       'FR_WIND', 'DE_TEMP', 'FR_TEMP', 'GAS_RET', 'COAL_RET', 'CARBON_RET'],
      dtype='object')

In [220]:
x_train.columns

Index(['ID', 'DAY_ID', 'COUNTRY', 'DE_CONSUMPTION', 'FR_CONSUMPTION',
       'DE_FR_EXCHANGE', 'FR_DE_EXCHANGE', 'DE_NET_EXPORT', 'FR_NET_EXPORT',
       'DE_NET_IMPORT', 'FR_NET_IMPORT', 'DE_GAS', 'FR_GAS', 'DE_COAL',
       'FR_COAL', 'DE_HYDRO', 'FR_HYDRO', 'DE_NUCLEAR', 'FR_NUCLEAR',
       'DE_SOLAR', 'FR_SOLAR', 'DE_WINDPOW', 'FR_WINDPOW', 'DE_LIGNITE',
       'DE_RESIDUAL_LOAD', 'FR_RESIDUAL_LOAD', 'DE_RAIN', 'FR_RAIN', 'DE_WIND',
       'FR_WIND', 'DE_TEMP', 'FR_TEMP', 'GAS_RET', 'COAL_RET', 'CARBON_RET'],
      dtype='object')

In [221]:
y_train.columns

Index(['ID', 'TARGET'], dtype='object')

In [222]:
y_test_random_final.columns

Index(['ID', 'TARGET'], dtype='object')

In [223]:
y_test_random_final.size

1308

# DATA PREPROCESSING

In [224]:
x_train_encoded = pd.get_dummies(x_train, columns = ['COUNTRY'])


In [225]:
x_test_encoded = pd.get_dummies(x_test_final, columns = ['COUNTRY'])


In [226]:
#x_train_encoded.head()

#### Imputing Missing Values with Median

In [227]:
x_train_clean = x_train_encoded.fillna(x_train_encoded.median())


In [228]:
x_test_clean = x_test_encoded.fillna(x_test_encoded.median())


In [229]:
x_train_clean

ID  DAY_ID  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
0     1054     206        0.210099       -0.427458       -0.606523   
1     2049     501       -0.022399       -1.003452       -0.022063   
2     1924     687        1.395035        1.978665        1.021305   
3      297     720       -0.983324       -0.849198       -0.839586   
4     1101     818        0.143807       -0.617038       -0.924990   
...    ...     ...             ...             ...             ...   
1489   459     809        1.529204        1.106682       -1.855327   
1490  1674     887        1.618582        1.752840        0.611392   
1491   748    1083        0.856399        0.489199       -0.255778   
1492  1454    1133        0.560689       -0.343777       -0.830239   
1493   359    1118        0.226730       -0.711005        0.873016   

      FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  \
0           0.606523      -0.306899       0.692860       0.306899   
1           0.022063      -0.573520      -1.130838       0.573520   
2          -1.021305      -0.622021      -1.682587       0.622021   
3           0.839586      -0.270870       0.563230       0.270870   
4           0.924990      -0.306899       0.990324       0.306899   
...              ...            ...            ...            ...   
1489        1.855327      -0.218658       1.450426       0.218658   
1490       -0.611392       0.449153      -0.152146      -0.449153   
1491        0.255778      -1.531544      -0.829568       1.531544   
1492        0.830239      -0.304856       1.210230       0.304856   
1493       -0.873016       0.296651      -1.547075      -0.296651   

      FR_NET_IMPORT  ...   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP   FR_TEMP  \
0         -0.692860  ... -0.556356 -0.790823 -0.283160 -1.069070 -0.063404   
1          1.130838  ... -0.770457  1.522331  0.828412  0.437419  1.831241   
2          1.682587  ... -0.313338  0.431134  0.487608  0.684884  0.114836   
3         -0.563230  ... -0.507570 -0.499409 -0.236249  0.350938 -0.417514   
4         -0.990324  ... -0.424550 -1.088158 -1.011560  0.614338  0.729495   
...             ...  ...       ...       ...       ...       ...       ...   
1489      -1.450426  ... -0.228147 -0.261571 -0.229031 -0.026306 -0.020889   
1490       0.152146  ... -0.228147 -0.261571 -0.229031 -0.026306 -0.020889   
1491       0.829568  ...  0.404763 -0.594595  0.894011  0.256338  0.402316   
1492      -1.210230  ... -0.390304 -0.972088 -1.501930  1.215528  1.338708   
1493       1.547075  ... -0.744355 -0.482286 -0.941070 -0.120703  0.391491   

       GAS_RET  COAL_RET  CARBON_RET  COUNTRY_DE  COUNTRY_FR  
0     0.339041  0.124552   -0.002445           0           1  
1    -0.659091  0.047114   -0.490365           0           1  
2     0.535974  0.743338    0.204952           0           1  
3     0.911652 -0.296168    1.073948           1           0  
4     0.245109  1.526606    2.614378           0           1  
...        ...       ...         ...         ...         ...  
1489  0.876984  0.819520    1.320373           1           0  
1490  0.932633 -0.085690    0.356356           0           1  
1491 -1.112899 -0.237835    0.067152           1           0  
1492  0.962812 -5.392852   -0.843812           0           1  
1493  1.059829 -0.326640   -0.566724           1           0  

[1494 rows x 36 columns]

In [230]:
y_train

ID    TARGET
0     1054  0.028313
1     2049 -0.112516
2     1924 -0.180840
3      297 -0.260356
4     1101 -0.071733
...    ...       ...
1489   459 -0.172597
1490  1674 -0.063546
1491   748  0.151797
1492  1454 -0.640917
1493   359 -0.252764

[1494 rows x 2 columns]

# MUTUAL INFORMATION

Mutual information is a lot like correlation in that it measures a relationship between two quantities. The advantage of mutual information is that it can detect any kind of relationship, while correlation only detects linear relationships.



In [231]:
# For finding mutual information all feature data must be treated as discrete and hence shall be in "numeric" format

discrete_features = x_train_clean.dtypes

In [232]:
discrete_features

ID                    int64
DAY_ID                int64
DE_CONSUMPTION      float64
FR_CONSUMPTION      float64
DE_FR_EXCHANGE      float64
FR_DE_EXCHANGE      float64
DE_NET_EXPORT       float64
FR_NET_EXPORT       float64
DE_NET_IMPORT       float64
FR_NET_IMPORT       float64
DE_GAS              float64
FR_GAS              float64
DE_COAL             float64
FR_COAL             float64
DE_HYDRO            float64
FR_HYDRO            float64
DE_NUCLEAR          float64
FR_NUCLEAR          float64
DE_SOLAR            float64
FR_SOLAR            float64
DE_WINDPOW          float64
FR_WINDPOW          float64
DE_LIGNITE          float64
DE_RESIDUAL_LOAD    float64
FR_RESIDUAL_LOAD    float64
DE_RAIN             float64
FR_RAIN             float64
DE_WIND             float64
FR_WIND             float64
DE_TEMP             float64
FR_TEMP             float64
GAS_RET             float64
COAL_RET            float64
CARBON_RET          float64
COUNTRY_DE            uint8
COUNTRY_FR          

In [233]:
x_train_clean

ID  DAY_ID  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
0     1054     206        0.210099       -0.427458       -0.606523   
1     2049     501       -0.022399       -1.003452       -0.022063   
2     1924     687        1.395035        1.978665        1.021305   
3      297     720       -0.983324       -0.849198       -0.839586   
4     1101     818        0.143807       -0.617038       -0.924990   
...    ...     ...             ...             ...             ...   
1489   459     809        1.529204        1.106682       -1.855327   
1490  1674     887        1.618582        1.752840        0.611392   
1491   748    1083        0.856399        0.489199       -0.255778   
1492  1454    1133        0.560689       -0.343777       -0.830239   
1493   359    1118        0.226730       -0.711005        0.873016   

      FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  \
0           0.606523      -0.306899       0.692860       0.306899   
1           0.022063      -0.573520      -1.130838       0.573520   
2          -1.021305      -0.622021      -1.682587       0.622021   
3           0.839586      -0.270870       0.563230       0.270870   
4           0.924990      -0.306899       0.990324       0.306899   
...              ...            ...            ...            ...   
1489        1.855327      -0.218658       1.450426       0.218658   
1490       -0.611392       0.449153      -0.152146      -0.449153   
1491        0.255778      -1.531544      -0.829568       1.531544   
1492        0.830239      -0.304856       1.210230       0.304856   
1493       -0.873016       0.296651      -1.547075      -0.296651   

      FR_NET_IMPORT  ...   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP   FR_TEMP  \
0         -0.692860  ... -0.556356 -0.790823 -0.283160 -1.069070 -0.063404   
1          1.130838  ... -0.770457  1.522331  0.828412  0.437419  1.831241   
2          1.682587  ... -0.313338  0.431134  0.487608  0.684884  0.114836   
3         -0.563230  ... -0.507570 -0.499409 -0.236249  0.350938 -0.417514   
4         -0.990324  ... -0.424550 -1.088158 -1.011560  0.614338  0.729495   
...             ...  ...       ...       ...       ...       ...       ...   
1489      -1.450426  ... -0.228147 -0.261571 -0.229031 -0.026306 -0.020889   
1490       0.152146  ... -0.228147 -0.261571 -0.229031 -0.026306 -0.020889   
1491       0.829568  ...  0.404763 -0.594595  0.894011  0.256338  0.402316   
1492      -1.210230  ... -0.390304 -0.972088 -1.501930  1.215528  1.338708   
1493       1.547075  ... -0.744355 -0.482286 -0.941070 -0.120703  0.391491   

       GAS_RET  COAL_RET  CARBON_RET  COUNTRY_DE  COUNTRY_FR  
0     0.339041  0.124552   -0.002445           0           1  
1    -0.659091  0.047114   -0.490365           0           1  
2     0.535974  0.743338    0.204952           0           1  
3     0.911652 -0.296168    1.073948           1           0  
4     0.245109  1.526606    2.614378           0           1  
...        ...       ...         ...         ...         ...  
1489  0.876984  0.819520    1.320373           1           0  
1490  0.932633 -0.085690    0.356356           0           1  
1491 -1.112899 -0.237835    0.067152           1           0  
1492  0.962812 -5.392852   -0.843812           0           1  
1493  1.059829 -0.326640   -0.566724           1           0  

[1494 rows x 36 columns]

In [234]:
x_train_clean_modified = x_train_clean.iloc[:, 2:] # Without ID and Date ID


In [235]:
x_train_clean_modified

DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  FR_DE_EXCHANGE  \
0           0.210099       -0.427458       -0.606523        0.606523   
1          -0.022399       -1.003452       -0.022063        0.022063   
2           1.395035        1.978665        1.021305       -1.021305   
3          -0.983324       -0.849198       -0.839586        0.839586   
4           0.143807       -0.617038       -0.924990        0.924990   
...              ...             ...             ...             ...   
1489        1.529204        1.106682       -1.855327        1.855327   
1490        1.618582        1.752840        0.611392       -0.611392   
1491        0.856399        0.489199       -0.255778        0.255778   
1492        0.560689       -0.343777       -0.830239        0.830239   
1493        0.226730       -0.711005        0.873016       -0.873016   

      DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  FR_NET_IMPORT    DE_GAS  \
0         -0.306899       0.692860       0.306899      -0.692860  0.441238   
1         -0.573520      -1.130838       0.573520       1.130838  0.174773   
2         -0.622021      -1.682587       0.622021       1.682587  2.351913   
3         -0.270870       0.563230       0.270870      -0.563230  0.487818   
4         -0.306899       0.990324       0.306899      -0.990324  0.238693   
...             ...            ...            ...            ...       ...   
1489      -0.218658       1.450426       0.218658      -1.450426  1.810665   
1490       0.449153      -0.152146      -0.449153       0.152146  1.972779   
1491      -1.531544      -0.829568       1.531544       0.829568  2.108764   
1492      -0.304856       1.210230       0.304856      -1.210230 -0.003973   
1493       0.296651      -1.547075      -0.296651       1.547075  0.452981   

        FR_GAS  ...   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP   FR_TEMP  \
0    -0.213766  ... -0.556356 -0.790823 -0.283160 -1.069070 -0.063404   
1     0.426940  ... -0.770457  1.522331  0.828412  0.437419  1.831241   
2     2.122241  ... -0.313338  0.431134  0.487608  0.684884  0.114836   
3     0.194659  ... -0.507570 -0.499409 -0.236249  0.350938 -0.417514   
4    -0.240862  ... -0.424550 -1.088158 -1.011560  0.614338  0.729495   
...        ...  ...       ...       ...       ...       ...       ...   
1489  1.388269  ... -0.228147 -0.261571 -0.229031 -0.026306 -0.020889   
1490  1.558300  ... -0.228147 -0.261571 -0.229031 -0.026306 -0.020889   
1491  1.866399  ...  0.404763 -0.594595  0.894011  0.256338  0.402316   
1492  0.869742  ... -0.390304 -0.972088 -1.501930  1.215528  1.338708   
1493  0.171967  ... -0.744355 -0.482286 -0.941070 -0.120703  0.391491   

       GAS_RET  COAL_RET  CARBON_RET  COUNTRY_DE  COUNTRY_FR  
0     0.339041  0.124552   -0.002445           0           1  
1    -0.659091  0.047114   -0.490365           0           1  
2     0.535974  0.743338    0.204952           0           1  
3     0.911652 -0.296168    1.073948           1           0  
4     0.245109  1.526606    2.614378           0           1  
...        ...       ...         ...         ...         ...  
1489  0.876984  0.819520    1.320373           1           0  
1490  0.932633 -0.085690    0.356356           0           1  
1491 -1.112899 -0.237835    0.067152           1           0  
1492  0.962812 -5.392852   -0.843812           0           1  
1493  1.059829 -0.326640   -0.566724           1           0  

[1494 rows x 34 columns]

In [236]:
y_train_modified = y_train.iloc[:,1:] #Without ID
y_train_modified

TARGET
0     0.028313
1    -0.112516
2    -0.180840
3    -0.260356
4    -0.071733
...        ...
1489 -0.172597
1490 -0.063546
1491  0.151797
1492 -0.640917
1493 -0.252764

[1494 rows x 1 columns]

In [237]:
"""from sklearn.feature_selection import mutual_info_regression

def make_mi_scores(X, y, discrete_features):
    mi_scores = mutual_info_regression(X, y, discrete_features=discrete_features)
    mi_scores = pd.Series(mi_scores, name="MI Scores", index=X.columns)
    mi_scores = mi_scores.sort_values(ascending=False)
    return mi_scores

mi_scores = make_mi_scores(x_train_clean_modified, y_train_modified, discrete_features)
mi_scores[::]  # show a few features with their MI scores"""

'from sklearn.feature_selection import mutual_info_regression\n\ndef make_mi_scores(X, y, discrete_features):\n    mi_scores = mutual_info_regression(X, y, discrete_features=discrete_features)\n    mi_scores = pd.Series(mi_scores, name="MI Scores", index=X.columns)\n    mi_scores = mi_scores.sort_values(ascending=False)\n    return mi_scores\n\nmi_scores = make_mi_scores(x_train_clean_modified, y_train_modified, discrete_features)\nmi_scores[::]  # show a few features with their MI scores'

In [238]:
"""def plot_mi_scores(scores):
    scores = scores.sort_values(ascending=True)
    width = np.arange(len(scores))
    ticks = list(scores.index)
    plt.barh(width, scores)
    plt.yticks(width, ticks)
    plt.title("Mutual Information Scores")


plt.figure(dpi=100, figsize=(9, 6))
plot_mi_scores(mi_scores)"""

'def plot_mi_scores(scores):\n    scores = scores.sort_values(ascending=True)\n    width = np.arange(len(scores))\n    ticks = list(scores.index)\n    plt.barh(width, scores)\n    plt.yticks(width, ticks)\n    plt.title("Mutual Information Scores")\n\n\nplt.figure(dpi=100, figsize=(9, 6))\nplot_mi_scores(mi_scores)'

# Feature Engineering

- Energy Production Efficiency: DE_WIND / DE_CONSUMPTION or FR_SOLAR / FR_CONSUMPTION to capture how much of the consumption is covered by renewable sources.

- Weather Impact on Energy: FR_RAIN * FR_HYDRO to see if rain has a direct impact on hydroelectric power production.

- Net Exchange Ratios: Compute the net exchange ratios between countries to understand the dependency or contribution of cross-border exchanges to the total energy consumption or production, e.g., (DE_FR_EXCHANGE - FR_DE_EXCHANGE) / (DE_CONSUMPTION + FR_CONSUMPTION).

- Cross-Feature Interactions: (FR_WIND + FR_SOLAR) / FR_CONSUMPTION to capture the proportion of renewable energy in total consumption.

- Temperature Effect on Consumption: Since temperature can influence energy demand (e.g., heating in winter, cooling in summer), consider creating features like FR_TEMP * FR_CONSUMPTION and DE_TEMP * DE_CONSUMPTION to model this relationship.

- Renewable vs. Non-renewable Ratios: Calculate the ratio of renewable to non-renewable energy production for each country, e.g., (FR_WINDPOW + FR_SOLAR) / (FR_COAL + FR_GAS).

- Carbon Intensity: Create a feature representing the carbon intensity of electricity generation, e.g., (COAL_RET + GAS_RET) / (DE_CONSUMPTION + FR_CONSUMPTION), assuming coal and gas have higher carbon emissions compared to renewables.

In [239]:
def feature_engineering(df):
  df['energy_wind'] = df['DE_WIND']/df['DE_CONSUMPTION']
  df['energy_solar'] = df['FR_SOLAR']/df['FR_CONSUMPTION']
  df['weather_impact'] = df['FR_RAIN']*df['FR_HYDRO']
  df['net_energy_ratio'] = df['FR_DE_EXCHANGE'] - df['DE_FR_EXCHANGE'] / (df['DE_CONSUMPTION']+df['FR_CONSUMPTION'])
  df['cross_feature'] = (df['FR_WIND']+df['FR_SOLAR'])/df['FR_CONSUMPTION']
  df['FR_temp_impact'] = df['FR_TEMP']+df['FR_CONSUMPTION']
  df['DE_temp_impact'] = df['DE_TEMP']+df['DE_CONSUMPTION']
  df['REN_NON'] = (df['FR_WINDPOW'] + df['FR_SOLAR']) / (df['FR_COAL'] + df['FR_GAS'])
  df['carbon_intensity'] = (df['COAL_RET'] + df['GAS_RET']) / (df['DE_CONSUMPTION'] + df['FR_CONSUMPTION'])
  return df

In [240]:
X = feature_engineering(x_train_clean_modified)

In [241]:
X.columns

Index(['DE_CONSUMPTION', 'FR_CONSUMPTION', 'DE_FR_EXCHANGE', 'FR_DE_EXCHANGE',
       'DE_NET_EXPORT', 'FR_NET_EXPORT', 'DE_NET_IMPORT', 'FR_NET_IMPORT',
       'DE_GAS', 'FR_GAS', 'DE_COAL', 'FR_COAL', 'DE_HYDRO', 'FR_HYDRO',
       'DE_NUCLEAR', 'FR_NUCLEAR', 'DE_SOLAR', 'FR_SOLAR', 'DE_WINDPOW',
       'FR_WINDPOW', 'DE_LIGNITE', 'DE_RESIDUAL_LOAD', 'FR_RESIDUAL_LOAD',
       'DE_RAIN', 'FR_RAIN', 'DE_WIND', 'FR_WIND', 'DE_TEMP', 'FR_TEMP',
       'GAS_RET', 'COAL_RET', 'CARBON_RET', 'COUNTRY_DE', 'COUNTRY_FR',
       'energy_wind', 'energy_solar', 'weather_impact', 'net_energy_ratio',
       'cross_feature', 'FR_temp_impact', 'DE_temp_impact', 'REN_NON',
       'carbon_intensity'],
      dtype='object')

In [242]:
X.head()

DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  FR_DE_EXCHANGE  \
0        0.210099       -0.427458       -0.606523        0.606523   
1       -0.022399       -1.003452       -0.022063        0.022063   
2        1.395035        1.978665        1.021305       -1.021305   
3       -0.983324       -0.849198       -0.839586        0.839586   
4        0.143807       -0.617038       -0.924990        0.924990   

   DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  FR_NET_IMPORT    DE_GAS  \
0      -0.306899       0.692860       0.306899      -0.692860  0.441238   
1      -0.573520      -1.130838       0.573520       1.130838  0.174773   
2      -0.622021      -1.682587       0.622021       1.682587  2.351913   
3      -0.270870       0.563230       0.270870      -0.563230  0.487818   
4      -0.306899       0.990324       0.306899      -0.990324  0.238693   

     FR_GAS  ...  COUNTRY_FR  energy_wind  energy_solar  weather_impact  \
0 -0.213766  ...           1    -3.764056     -2.921715       -0.115632   
1  0.426940  ...           1   -67.963185     -3.226242        0.621845   
2  2.122241  ...           1     0.309049     -0.187520       -0.244135   
3  0.194659  ...           0     0.507879     -0.662913       -0.670238   
4 -0.240862  ...           1    -7.566776     -1.172680        0.337934   

   net_energy_ratio  cross_feature  FR_temp_impact  DE_temp_impact    REN_NON  \
0         -2.183893      -2.259287       -0.490862       -0.858971  13.056610   
1          0.000556      -4.051804        0.827788        0.415020  -9.337436   
2         -1.324031       0.058913        2.093502        2.079919  -0.176691   
3          0.381427      -0.384711       -1.266712       -0.632386  -1.572342   
4         -1.029640       0.466701        0.112457        0.758146  -0.308411   

   carbon_intensity  
0         -2.132838  
1          0.596555  
2          0.379201  
3         -0.335867  
4         -3.743874  

[5 rows x 43 columns]

#### IMPROVING BASELINE BENCHMARKING WITH NEW FEATURES

In [243]:
import numpy as np
import pandas as pd
from scipy.stats import spearmanr
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor


In [244]:
lr = LinearRegression()
X.columns

Index(['DE_CONSUMPTION', 'FR_CONSUMPTION', 'DE_FR_EXCHANGE', 'FR_DE_EXCHANGE',
       'DE_NET_EXPORT', 'FR_NET_EXPORT', 'DE_NET_IMPORT', 'FR_NET_IMPORT',
       'DE_GAS', 'FR_GAS', 'DE_COAL', 'FR_COAL', 'DE_HYDRO', 'FR_HYDRO',
       'DE_NUCLEAR', 'FR_NUCLEAR', 'DE_SOLAR', 'FR_SOLAR', 'DE_WINDPOW',
       'FR_WINDPOW', 'DE_LIGNITE', 'DE_RESIDUAL_LOAD', 'FR_RESIDUAL_LOAD',
       'DE_RAIN', 'FR_RAIN', 'DE_WIND', 'FR_WIND', 'DE_TEMP', 'FR_TEMP',
       'GAS_RET', 'COAL_RET', 'CARBON_RET', 'COUNTRY_DE', 'COUNTRY_FR',
       'energy_wind', 'energy_solar', 'weather_impact', 'net_energy_ratio',
       'cross_feature', 'FR_temp_impact', 'DE_temp_impact', 'REN_NON',
       'carbon_intensity'],
      dtype='object')

In [245]:
X = X.drop(columns = ['COUNTRY_DE','COUNTRY_FR'],axis = 1)

In [246]:
lr.fit(X, y_train_modified)

output_train = lr.predict(X)

def metric_train(output):

    return  spearmanr(output, y_train_modified).correlation

print('Spearman correlation for the train set: {:.1f}%'.format(100 * metric_train(output_train) ))


Spearman correlation for the train set: 28.3%


From above we can observe that we increased the spearman correlation score from 27.9% previously to 28.3% by adding new features

In [247]:
# Copy ID column for submission
Y_test_submission = x_test_final[['ID']].copy()

# Fill missing values with median and encode categorical variables
x_test_clean = x_test_final.fillna(x_test_final.median())
x_encoded = pd.get_dummies(x_test_clean, columns=['COUNTRY'])

# Apply custom feature engineering (implementation not shown)
x_test_clean_modified = feature_engineering(x_encoded)

# Drop unnecessary columns from the modified DataFrame
x_test_clean_modified = x_test_clean_modified.drop(columns=['DAY_ID', 'ID'], axis=1)

# Assuming you want to drop 'COUNTRY_DE' and 'COUNTRY_FR' after encoding
x_test_clean_modified = x_test_clean_modified.drop(columns=['COUNTRY_DE', 'COUNTRY_FR'], axis=1)


<ipython-input-247-6e0a6833c569>:5: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  x_test_clean = x_test_final.fillna(x_test_final.median())


In [248]:
"""Y_test_submission = x_test_final[['ID']].copy()
x_test_clean = x_test_final.fillna(x_test_final.median())
x_encoded =  pd.get_dummies(x_test_clean, columns = ['COUNTRY'])

x_test_clean_modified = feature_engineering(x_encoded)
x = x_test_encoded.drop(columns = ['DAY_ID','ID'],axis = 1)
x = x_test_encoded.drop(columns = ['COUNTRY_DE','COUNTRY_FR'],axis = 1)"""

"Y_test_submission = x_test_final[['ID']].copy()\nx_test_clean = x_test_final.fillna(x_test_final.median())\nx_encoded =  pd.get_dummies(x_test_clean, columns = ['COUNTRY'])\n\nx_test_clean_modified = feature_engineering(x_encoded)\nx = x_test_encoded.drop(columns = ['DAY_ID','ID'],axis = 1)\nx = x_test_encoded.drop(columns = ['COUNTRY_DE','COUNTRY_FR'],axis = 1)"

In [249]:
x

ID  DAY_ID  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
0    1115     241        0.340083       -0.433604       -0.423521   
1    1202    1214        0.803209        0.780411        0.601610   
2    1194    1047        0.795540        0.721954        1.179158   
3    1084    1139        0.172555       -0.723427       -0.044539   
4    1135     842        0.949714        0.420236        0.617391   
..    ...     ...             ...             ...             ...   
649   879     213        0.267684       -0.901332        0.749769   
650   673     943        1.389207        1.466549        0.661778   
651  1641     261        1.377650        1.781671        0.557006   
652   712    1082        1.232202        1.971074        0.537247   
653  1060     889        0.152683       -0.436453       -0.014912   

     FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  \
0          0.423521       0.165333       0.519419      -0.165333   
1         -0.601610       0.342802       0.555367      -0.342802   
2         -1.179158       1.620928       0.666901      -1.620928   
3          0.044539            NaN      -0.205276            NaN   
4         -0.617391       0.608561      -0.240856      -0.608561   
..              ...            ...            ...            ...   
649       -0.749769       1.198565      -1.614458      -1.198565   
650       -0.661778       0.610986      -1.934012      -0.610986   
651       -0.557006      -0.326698      -1.462676       0.326698   
652       -0.537247      -1.109420      -1.464287       1.109420   
653        0.014912            NaN       0.827405            NaN   

     FR_NET_IMPORT  ...  FR_RESIDUAL_LOAD   DE_RAIN   FR_RAIN   DE_WIND  \
0        -0.519419  ...         -0.222525 -0.513180 -0.182048 -0.982546   
1        -0.555367  ...          0.857739 -0.340595 -0.301094 -0.759816   
2        -0.666901  ...          0.447967  0.796475 -0.367248  0.376055   
3         0.205276  ...         -0.561295 -0.542606 -0.013291 -0.791119   
4         0.240856  ...          0.503567 -0.230291 -0.609203 -0.744986   
..             ...  ...               ...       ...       ...       ...   
649       1.614458  ...         -1.057186  1.591562 -0.284628  2.272121   
650       1.934012  ...          1.055349  0.155454 -0.168436  1.290495   
651       1.462676  ...          1.824478 -0.825759 -0.266881 -1.247165   
652       1.464287  ...          1.964292 -0.145241 -0.358861  0.195383   
653      -0.827405  ...         -0.315320 -0.673333 -0.013776 -0.312564   

      FR_WIND   DE_TEMP   FR_TEMP   GAS_RET  COAL_RET  CARBON_RET  
0   -0.876632  0.880491  0.692242  0.569419 -0.029697   -0.929256  
1   -1.221443 -0.616617 -0.737496  0.251251  0.753646    0.664086  
2   -0.483363  0.865138  0.120079 -1.485642 -0.326450   -0.349747  
3   -0.894309  0.239153  0.457457 -0.746863  2.262654    0.642069  
4   -1.196282  0.176557  0.312557 -2.219626 -0.509272   -0.488341  
..        ...       ...       ...       ...       ...         ...  
649  1.647899 -0.838634 -0.463684 -1.071829  0.720641   -1.842072  
650  2.199105 -0.329098 -1.172309 -1.290355 -0.897009   -0.925825  
651 -0.666308 -0.807636 -1.260777  0.931572  0.151169    0.474347  
652 -0.103433  1.352906 -0.485161  0.995363 -0.152287    2.067306  
653 -0.873435  0.210927 -0.323090  2.143341  0.834252    0.787723  

[654 rows x 34 columns]

In [250]:
Y_test_submission['TARGET'] = lr.predict(x_test_clean_modified)


In [251]:
Y_test_submission.to_csv('benchmark_baseline_1.csv', index=False)


In [252]:
x_test_clean_modified.columns

Index(['DE_CONSUMPTION', 'FR_CONSUMPTION', 'DE_FR_EXCHANGE', 'FR_DE_EXCHANGE',
       'DE_NET_EXPORT', 'FR_NET_EXPORT', 'DE_NET_IMPORT', 'FR_NET_IMPORT',
       'DE_GAS', 'FR_GAS', 'DE_COAL', 'FR_COAL', 'DE_HYDRO', 'FR_HYDRO',
       'DE_NUCLEAR', 'FR_NUCLEAR', 'DE_SOLAR', 'FR_SOLAR', 'DE_WINDPOW',
       'FR_WINDPOW', 'DE_LIGNITE', 'DE_RESIDUAL_LOAD', 'FR_RESIDUAL_LOAD',
       'DE_RAIN', 'FR_RAIN', 'DE_WIND', 'FR_WIND', 'DE_TEMP', 'FR_TEMP',
       'GAS_RET', 'COAL_RET', 'CARBON_RET', 'energy_wind', 'energy_solar',
       'weather_impact', 'net_energy_ratio', 'cross_feature', 'FR_temp_impact',
       'DE_temp_impact', 'REN_NON', 'carbon_intensity'],
      dtype='object')

In [253]:
X.columns

Index(['DE_CONSUMPTION', 'FR_CONSUMPTION', 'DE_FR_EXCHANGE', 'FR_DE_EXCHANGE',
       'DE_NET_EXPORT', 'FR_NET_EXPORT', 'DE_NET_IMPORT', 'FR_NET_IMPORT',
       'DE_GAS', 'FR_GAS', 'DE_COAL', 'FR_COAL', 'DE_HYDRO', 'FR_HYDRO',
       'DE_NUCLEAR', 'FR_NUCLEAR', 'DE_SOLAR', 'FR_SOLAR', 'DE_WINDPOW',
       'FR_WINDPOW', 'DE_LIGNITE', 'DE_RESIDUAL_LOAD', 'FR_RESIDUAL_LOAD',
       'DE_RAIN', 'FR_RAIN', 'DE_WIND', 'FR_WIND', 'DE_TEMP', 'FR_TEMP',
       'GAS_RET', 'COAL_RET', 'CARBON_RET', 'energy_wind', 'energy_solar',
       'weather_impact', 'net_energy_ratio', 'cross_feature', 'FR_temp_impact',
       'DE_temp_impact', 'REN_NON', 'carbon_intensity'],
      dtype='object')